bundles/org.openhab.core.io.http.auth/src/main/java/org/openhab/core/io/http/auth/internal/AbstractAuthPageServlet.java

```diff
@@ -0,0 +1,129 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.core.io.http.auth.internal;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletResponse;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.core.auth.Authentication;
+import org.openhab.core.auth.AuthenticationException;
+import org.openhab.core.auth.AuthenticationProvider;
+import org.openhab.core.auth.User;
+import org.openhab.core.auth.UserRegistry;
+import org.openhab.core.auth.UsernamePasswordCredentials;
+import org.osgi.framework.BundleContext;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.http.HttpService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Abstract class for servlets to perform sensible operations requiring user authentication.
+ *
+ * @author Yannick Schaus - initial contribution
+ *
+ */
+@NonNullByDefault
+public abstract class AbstractAuthPageServlet extends HttpServlet {
+
+    protected static final long serialVersionUID = 5340598701104679840L;
+
+    private final Logger logger = LoggerFactory.getLogger(AbstractAuthPageServlet.class);
+
+    protected HttpService httpService;
+    protected UserRegistry userRegistry;
+    protected AuthenticationProvider authProvider;
+    @Nullable
+    protected Instant lastAuthenticationFailure;
+    protected int authenticationFailureCount = 0;
+
+    protected HashMap<String, Instant> csrfTokens = new HashMap<>();
+
+    protected String pageTemplate;
+
+    public AbstractAuthPageServlet(BundleContext bundleContext, @Reference HttpService httpService,
+            @Reference UserRegistry userRegistry, @Reference AuthenticationProvider authProvider) {
+        this.httpService = httpService;
+        this.userRegistry = userRegistry;
+        this.authProvider = authProvider;
+
+        pageTemplate = "";
+        URL resource = bundleContext.getBundle().getResource("pages/authorize.html");
+        if (resource != null) {
+            try {
+                pageTemplate = new String(resource.openStream().readAllBytes(), StandardCharsets.UTF_8);
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
```

Comment:
This is only when the page cannot be loaded from the bundle resources, I don't think it will ever happen. I didn't know how to handle this exception, any suggestion is welcome.
In general authentication failures are logged but not shown in detail to the user on the webpage (it only says "Please try again") because you don't want to divulge too much (i.e. does the account exist, and only the password is wrong, etc.).

Reply:
You'll run into this whenever one day the file is deleted/renamed or moved to another directory.
So I would add a message to the exception that it cannot open "pages/authorize.html".